operations/terraform-state.yaml

```diff
@@ -0,0 +1,111 @@
+---
+# Copyright 2020 widdix GmbH
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+AWSTemplateFormatVersion: '2010-09-09'
+Description: 'Terraform state (S3 and DynamoDB)'
+Metadata:
+  'AWS::CloudFormation::Interface':
+    ParameterGroups:
+    - Label:
+        default: 'Parent Stacks'
+      Parameters:
+      - ParentKmsKeyStack
+    - Label:
+        default: 'Terraform State Parameters'
+      Parameters:
+      - TerrformStateIdentifier
+      - TerraformStateUserARNs
+      - TerraformStateAdminARNs
+Parameters:
+  ParentKmsKeyStack:
+    Description: 'Stack name of parent KMS key stack based on security/kms-key.yaml template.'
+    Type: String
+  TerrformStateIdentifier:
+    Description: 'Globally unique identifier for your Terraform state.'
+    Type: String
+  TerraformStateUserARNs:
+    Description: 'A comma delimited list of IAM users or role ARNs that need to use the Terraform state (see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_notprincipal.html#specifying-notprincipal).'
+    Type: String
+  TerraformStateAdminARNs:
+    Description: 'A comma delimited list of IAM users or role ARNs that should be able to administer the Terraform state (see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_notprincipal.html#specifying-notprincipal).'
+    Type: String
+Resources:
+  StateBucket:
+    Type: AWS::S3::Bucket
+    Properties:
+      BucketEncryption:
+        ServerSideEncryptionConfiguration:
+        - ServerSideEncryptionByDefault:
+            KMSMasterKeyID: {'Fn::ImportValue': !Sub '${ParentKmsKeyStack}-KeyArn'}
+            SSEAlgorithm: 'aws:kms'
+      BucketName: !Ref TerrformStateIdentifier
+      PublicAccessBlockConfiguration:
+        BlockPublicAcls: true
+        BlockPublicPolicy: true
+        IgnorePublicAcls: true
+        RestrictPublicBuckets: true
+      VersioningConfiguration:
+        Status: 'Enabled'
+  StateBucketPolicy:
+    Type: AWS::S3::BucketPolicy
+    Properties:
+      Bucket: !Ref StateBucket
+      PolicyDocument:
+        Statement:
+        - Effect: 'Deny'
+          Action: 's3:*'
+          Resource:
+          - !Sub '${StateBucket.Arn}'
+          - !Sub '${StateBucket.Arn}/*'
+          NotPrincipal:
+            AWS: !Split [',', {"Fn::Sub": '${TerraformStateUserARNs},${TerraformStateAdminARNs}'}]
+        - Effect: 'Deny'
+          NotAction:
+          - 's3:ListBucket'
+          - 's3:GetObject'
```

Comment:
what about

* `s3:ListBucketVersions`
* `s3:GetObjectVersion`

Reply:
Terraform does not need access to those actions.

Reply:
I see